### PR TITLE
creating --exclude-if-present

### PR DESCRIPTION
**Pull request by idavydov**
Fixing #1808.

### CODE REVIEW

File: fs/filter.go [hunk 1]
```diff
@@ -385,6 +389,23 @@ func (f *Filter) includeRemote(remote string) bool {
...
 
+// ListContainsExcludeFile checks if exclude file is present in the list.
+func (f *Filter) ListContainsExcludeFile(entries DirEntries) bool {
+	if len(f.ExcludeFile) == 0 {
+		return false
+	}
+	for _, entry := range entries {
+		obj, ok := entry.(Object)
+		if ok {
+			_, leaf := RemoteSplit(obj.Remote())
```
Comment (ncw) [inline fs/filter.go]: I think you want `path.Base(obj.Remote())` here - `RemoteSplit` is for processing the `remote:path` urls.  Not very well named or documented, I'd agree!

**Comment by ncw** [comment]
Oh yes, and can you add some docs in `docs/content/filtering.md` - thanks!

**Comment by idavydov** [comment]
unfortunately the build for dcddb81 failed. but it's not a problem with the code. [buildserver could not access github.com during the build](https://travis-ci.org/ncw/rclone/jobs/299552367).

File: fs/walk.go [hunk 2]
```diff
@@ -250,6 +250,15 @@ func (dt DirTree) Dirs() (dirNames []string) {
 	return dirNames
 }
 
+// Prune remove directories from a directory tree.
+func (dt DirTree) Prune(dirNames []string) {
+	// We do not need to do a recursion.
+	// GC will take care of deleting child entries.
+	for _, dName := range dirNames {
+		delete(dt, dName)
```
Comment (ncw) [inline fs/walk.go]: `delete`-ing the entry is problematic as there will be a DirEntry of the deleted directory in the parent.

This leads to this with `rclone tree`

```
/
├── a
├── /sub [Couldn't find directory "sub"]
├── sub1
│   └── file1
└── sub2
    └── file2
```

I think setting the entry to nil (ie an empty slice) will work better

```
dt[dName] = nil
```

File: fs/walk.go [hunk 3]
```diff
@@ -250,6 +250,15 @@ func (dt DirTree) Dirs() (dirNames []string) {
 	return dirNames
 }
 
+// Prune remove directories from a directory tree.
+func (dt DirTree) Prune(dirNames []string) {
+	// We do not need to do a recursion.
+	// GC will take care of deleting child entries.
```
Comment (ncw) [inline fs/walk.go]: The child entries are still in the map though.  They will be GC-ed properly but I wonder if any uses of DirTree will mind...  I don't think they will at the moment.

**Comment by ncw** [comment]
> unfortunately the build for dcddb81 failed. but it's not a problem with the code

Yes there seems to be something up with the OS X builds today!

File: fs/walk.go [hunk 4]
```diff
@@ -250,13 +250,93 @@ func (dt DirTree) Dirs() (dirNames []string) {
...
+		parent := parentDir(dName)
+		dir, ok := dt[parent]
+		if !ok {
+			return errors.Errorf("parent directory of %s (%s) not in DirTree", dName, parent)
+		}
+		found := false
+		for i, entry := range dir {
+			switch x := entry.(type) {
+			case Directory:
+				if x.Remote() == dName {
```
Comment (idavydov) [inline fs/walk.go]: @ncw do we need to take a special care of case-insensitive fs? or in case-insensitive fs naming should be always consistent?

**Comment by idavydov** [comment]
Now `DirTree` is pruned properly. This might be a slower sometimes, but we are sure that `DirTree` is consistent. E.g. if we do not Prune properly, `DirTree.String()` will return unexpected results, which makes testing more difficult.
There is also one comment I added regarding case-insensitive filesystems. Do we expect something like this from them:
```
a/b
A/c
```
If this should not happen, the code should be more or less fine.
What do you think?

**Comment by ncw** [comment]
> Now DirTree is pruned properly. This might be a slower sometimes, but we are sure that DirTree is consistent. E.g. if we do not Prune properly, DirTree.String() will return unexpected results, which makes testing more difficult.

That looks good :-)

> There is also one comment I added regarding case-insensitive filesystems. Do we expect something like this from them:

No, we don't.  Rclone expects case insensitive file systems always to be consistent in the case presented, so your code looks fine.

I noticed one inconsistency between the `--fast-list` and non-fast list case

```
$ rclone --fast-list tree -a memstore:test-exclude-if-present
/
├── a
├── sub
│   ├── .ignore
│   ├── file
│   └── subsubdir
│       └── should-be-ignored
├── sub1
│   └── file1
└── sub2
    └── file2

4 directories, 6 files
```

```
$ rclone --exclude-if-present .ignore tree -a memstore:test-exclude-if-present
/
├── a
├── sub
├── sub1
│   └── file1
└── sub2
    └── file2

3 directories, 3 files
```

```
$ rclone --fast-list --exclude-if-present .ignore tree -a memstore:test-exclude-if-present
/
├── a
├── sub1
│   └── file1
└── sub2
    └── file2
```

Note the `sub` directory is missing in the `--fast-list` case.

I'd like the non-fast list and the fast list to be consistent here.  What do rdiff-backup and duplicity do?  Do they remove the directory completely or do they just leave an empty directory?

Leaving an empty directory is probably easiest to achieve as changing the non --fast-list case to remove the directory will be tricky.  I don't have a strong opinion on which option to choose, so I'll leave that to you!

**Comment by idavydov** [comment]
The proper way is to delete the directory (that's what `rdiff-backup` does). 

I had to update `IncludeDirectory`; now it returns a function (initialized by a `Fs` instance).

There are checks for exclude file in a few places:

- `IncludeDirectory`, when adding a new directory
- `ListDirSorted`, in case exclude file is present in the root of source directory
- `walkRDirTree`, as you mentioned a list can be in an arbitrary order

Also, I updated the documentation, and now it's written that `--exclude-if-present` has the highest priority over the other filters.

**Comment by idavydov** [comment]
Is it better like this?

**Comment by ncw** [comment]
Perfect - thank you for a great feature :-)

**Pull request merged**